File: app/miscellaneous/file/ChunkMaker.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package miscellaneous.file;

import org.apache.pekko.stream.Attributes;
import org.apache.pekko.stream.FlowShape;
import org.apache.pekko.stream.Inlet;
import org.apache.pekko.stream.Outlet;
import org.apache.pekko.stream.stage.AbstractInHandler;
import org.apache.pekko.stream.stage.AbstractOutHandler;
import org.apache.pekko.stream.stage.GraphStage;
import org.apache.pekko.stream.stage.GraphStageLogic;
import org.apache.pekko.util.ByteString;
import scala.Tuple2;

public class ChunkMaker extends GraphStage<FlowShape<ByteString, ByteString>> {

    private final int chunkSize;

    public Inlet<ByteString> in = Inlet.create("ChunkMaker.in");
    public Outlet<ByteString> out = Outlet.create("ChunkMaker.out");
    private final FlowShape<ByteString, ByteString> shape = FlowShape.of(in, out);

    public ChunkMaker(int chunkSize) {
        this.chunkSize = chunkSize;
    }

    @Override
    public FlowShape<ByteString, ByteString> shape() {
        return shape;
    }

    @Override
    public GraphStageLogic createLogic(Attributes inheritedAttributes) {
        return new GraphStageLogic(shape) {
            private ByteString buffer = ByteString.emptyByteString();

            {
                setHandler(
                    out,
                    new AbstractOutHandler() {
                        @Override
                        public void onPull() {
                            emitChunk();
                        }
                    }
                );

                setHandler(
                    in,
                    new AbstractInHandler() {
                        @Override
                        public void onPush() {
                            ByteString elem = grab(in);
                            buffer = buffer.concat(elem);
                            emitChunk();
                        }

                        @Override
                        public void onUpstreamFinish() {
                            if (buffer.isEmpty()) {
                                completeStage();
                                return;
                            }

                            if (isAvailable(out)) {
                                emitChunk();
                            }
                        }
                    }
                );
            }

            private void emitChunk() {
                if (buffer.isEmpty() && isClosed(in)) {
                    completeStage();
                    return;
                }
                if (buffer.size() < chunkSize && !isClosed(in)) {
                    pull(in);
                    return;
                }
                Tuple2<ByteString, ByteString> split = buffer.splitAt(chunkSize);
                ByteString chunk = split._1();
                buffer = split._2();
                push(out, chunk);
            }
        };
    }
}

File: app/services/file/ChunkMaker.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.file

import org.apache.pekko.stream._
import org.apache.pekko.stream.stage._
import org.apache.pekko.util.ByteString

/** GraphStage that chunks incoming ByteStrings into fixed-size chunks
  *
  * @param chunkSize
  *   the size of each chunk in bytes
  */
class ChunkMaker(chunkSize: Int) extends GraphStage[FlowShape[ByteString, ByteString]]:
  val in: Inlet[ByteString]   = Inlet("ChunkMaker.in")
  val out: Outlet[ByteString] = Outlet("ChunkMaker.out")

  override val shape: FlowShape[ByteString, ByteString] = FlowShape(in, out)

  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
    new GraphStageLogic(shape):
      private var buffer: ByteString = ByteString.empty

      setHandler(
        out,
        new OutHandler:
          override def onPull(): Unit = emitChunk()
      )

      setHandler(
        in,
        new InHandler:
          override def onPush(): Unit =
            val elem = grab(in)
            buffer = buffer.concat(elem)
            emitChunk()

          override def onUpstreamFinish(): Unit =
            if buffer.isEmpty then completeStage()
            else if isAvailable(out) then emitChunk()
      )

      private def emitChunk(): Unit =
        if buffer.isEmpty && isClosed(in) then completeStage()
        else if buffer.size < chunkSize && !isClosed(in) then pull(in)
        else
          val (chunk, remainder) = buffer.splitAt(chunkSize)
          buffer = remainder
          push(out, chunk)
